gef.py

```diff
@@ -5207,6 +5205,207 @@ def do_invoke(self, args: List) -> None:
+            return default
+
+    class Module:
+        def __init__(self, path: pathlib.Path) -> None:
+            self.path = path
+            self.name = path.stem
+            self.raw = self.__load()
+            self.structures : Dict[str, ExternalStructureManager.Structure] = {}
+
+            for entry in self:
+                structure = ExternalStructureManager.Structure(self.path, entry)
+                self.structures[structure.name] = structure
+            return
+
+        def __load(self) -> ModuleType:
+            """Load a custom module, and return it."""
+            fpath = self.path
+            spec = importlib.util.spec_from_file_location(fpath.stem, fpath)
+            module = importlib.util.module_from_spec(spec)
+            sys.modules[fpath.stem] = module
+            spec.loader.exec_module(module)
+            return module
+
+        def __str__(self) -> str:
+            return self.name
+
+        def __iter__(self) -> Generator[str, None, None]:
+            _invalid = {"BigEndianStructure", "LittleEndianStructure", "Structure"}
+            _structs = {x for x in dir(self.raw) \
+                             if inspect.isclass(getattr(self.raw, x)) \
+                             and issubclass(getattr(self.raw, x), ctypes.Structure)}
+            for entry in (_structs - _invalid):
+                yield entry
+            return
+
+        def __contains__(self, structure_name: str) -> bool:
+            return structure_name in self.structures
+
+        def __getitem__(self, structure_name: str) -> "ExternalStructureManager.Structure":
+            return self.structures[structure_name]
+
+    class Modules:
+        def __init__(self, path: pathlib.Path) -> None:
+            self.modules: Dict[str, ExternalStructureManager.Module] = {}
+            self.root: pathlib.Path = path
+
+            for entry in self.root.iterdir():
+                if not entry.is_file(): continue
+                if entry.suffix != ".py": continue
+                if entry.name == "__init__.py": continue
+                module = ExternalStructureManager.Module(entry)
+                self.modules[module.name] = module
+            return
+
+        def __contains__(self, structure_name: str) -> bool:
+            """Return True if the structure name is found in any of the modules"""
+            for module in self.modules:
+                if structure_name in module:
+                    return True
+            return False
+
+        def __getitem__(self, module_name: str) -> "ExternalStructureManager.Module":
+            return self.modules[module_name]
+
+        def __iter__(self) -> Iterator[str]:
+            return iter(self.modules)
+
+    def __init__(self):
+        self.clear_caches()
+        return
+
+    def clear_caches(self):
+        self._path = None
+        self._modules = None
+        return
+
+    @property
+    def modules(self) -> "ExternalStructureManager.Modules":
```

Comment:
So this means `ESM.modules` is of type `ESM.Modules`. `__getitem__` in that class makes it so that the way you iterate works, but why not just make this method return `self._modules.modules` which is a dictionary?

Reply:
This works nice now that `Modules` is a dict.